proposals/0000-bytearray-literals.rst

```diff
@@ -0,0 +1,345 @@
+
+The GHC's constant folding / rewrite mechanisms will be able to avoid creation
+of intermediate UTF8 encoded ``ByteArray#``.
+
+The conversion primops are also useful in non-static contexts.
+
+*Note:* this is not possible current ``Addr#`` as a ``String`` literal "backend".
+
+Fow now I don't propose to add recoding primops for ``Addr#``
+``Addr#`` is an arbitrary machine address outside the garbage-collected heap:
+how to create new ones? (AFAICS there aren't such primops currently).
+
+
+Effect and Interactions
+-----------------------
+
+Library authors may provide `RULES` to avoid looping to get the length of the string (``Prim.Builder``  in ``bytestring`` could simply ``memcpy`` the contents of a literal),
+or avoid roundtripping through `String` (also in ``bytestring``):
+
+.. code-block:: haskell
+
+  data ShortByteString = SBS ByteArray#
+ 
+  instance FromString ShortByteString where
+    fromString = packChars
+ 
+  {-# RULES packChars (unpackAsciiByteArray# ba) = SBS ba #-}
+  {-# RULES packChars (unpackUtf8ByteArray# ba) = SBS ba #-}
+
+  packChars :: ByteArray# -> ShortByteString
+  packChars = SBS
+
+This will resolve `Trac 5218 <https://ghc.haskell.org/trac/ghc/ticket/5218>`_.
+
+Similarly, for textual types (`compare with current <http://hackage.haskell.org/package/text-short-0.1.2/docs/src/Data.Text.Short.Internal.html#line-1455>`_)
+
+.. code-block:: haskell
+
+  -- short text in UTF-16
+  data ShortText16 = ST16 ByteArray#
+
+  {-# RULES "ShortText literal UTF-8" forall s . fromStringLit (GHC.unpackUtf8ByteArray# s) = fromLitMUtf8Addr# s #-}
+
+  fromLitUtf8Addr# :: ByteArray# -> ShortText
+  fromLitUtt8Addr# ba = ST (recodeUtf8ToUtf16# ba)
+
+Other effects and interactions are:
+
+- ``ByteArray#`` is a ``String#`` asked for in `Trac 11312 (comment 3) <https://ghc.haskell.org/trac/ghc/ticket/11312#comment:3>`_,
+  the only missing part is way to create statically allocated ``ByteArray#``. This proposal addresses that.
+
+- ``ByteArray#`` literals can be used to implement better ``mkInteger``. `Trac 9719: Improve mkInteger interface <https://ghc.haskell.org/trac/ghc/ticket/9719>`_.
+
+- This proposal will affect ``template-haskell``.
+  We'll need to (again) change the type of ``StringPrimL`` constuctor to include the result-type tag.
+  Tag type can also be used to implement encoding-tagged bytearray# for safer conversions
+  (`Unlifted newtypes <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0013-unlifted-newtypes.rst>`_ will allow that once implemented).
+
+Costs and Drawbacks
+-------------------
+
+This would be a potentially breaking change for any code using unboxed string literals.
+
+Alternatives
+------------
+
+One proposed alternative is to use double hash syntax ``"foo"##`` to represent
+UTF8 ``ByteArray#``.  That variant is very limited in power compared to proposed.
+
+In fact, we might introduce similar syntax for the number literals, e.g. ``120#i8``, if we get `Int8#` primitive type.
```

Comment:
I agree that the syntax that you propose above is *much* preferred to the "keep-adding-more-hashes" approach. I would really like to see the this syntax adopted for numeric llterals at some point as well.

Reply:
that'd be nice, agreed